src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,182 @@
+        T oldValue = node.item;
+        node.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkElementIndex(index);
+        return unlink(getExistingNodeByIndex(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (MyLinkedList.Node<T> node = first; node != null; node = node.next) {
+                if (node.item == null) {
+                    unlink(node);
+                    return true;
+                }
+            }
+        } else {
+            for (MyLinkedList.Node<T> node = first; node != null; node = node.next) {
+                if (object.equals(node.item)) {
+                    unlink(node);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    void linkLast(T node) {
+        final MyLinkedList.Node<T> lastNode = last;
+        final MyLinkedList.Node<T> newNode = new MyLinkedList.Node<>(last, node, null);
+        last = newNode;
+        if (lastNode == null) {
+            first = newNode;
+        } else {
+            lastNode.next = newNode;
+        }
+
+        size++;
+    }
+
+    void linkBefore(T value, Node<T> existingNode) {
+        final MyLinkedList.Node<T> previousNode = existingNode.prev;
+        final MyLinkedList.Node<T> newNode
+                = new MyLinkedList.Node<>(previousNode, value, existingNode);
+        existingNode.prev = newNode;
+        if (previousNode == null) {
+            first = newNode;
+        } else {
+            previousNode.next = newNode;
+        }
+        size++;
+    }
+
+    MyLinkedList.Node<T> getExistingNodeByIndex(int index) {
+        if (index < (size >> 1)) {
+            MyLinkedList.Node<T> node = first;
+            for (int i = 0; i < index; i++) {
+                node = node.next;
+            }
+            return node;
+        } else {
+            MyLinkedList.Node<T> node = last;
+            for (int i = size - 1; i > index; i--) {
+                node = node.prev;
+            }
+            return node;
+        }
+    }
+
+    private void checkPositionIndex(int index) {
+        if (!isPositionIndex(index)) {
+            throw new IndexOutOfBoundsException(outOfBoundsMsg(index));
+        }
+    }
+
+    private void checkElementIndex(int index) {
+        if (!isElementIndex(index)) {
+            throw new IndexOutOfBoundsException(outOfBoundsMsg(index));
+        }
+    }
+
+    private boolean isPositionIndex(int index) {
+        return index >= 0 && index <= size;
+    }
+
+    private boolean isElementIndex(int index) {
+        return index >= 0 && index < size;
+    }
```

Comment:
You use this booleans only once, perhaps it would be better to place it where it is used

Reply:
Disagree here, sorry